File: backend/src/bond_monitor/application/notifications/deliver_use_case.py
"""Deliver alerts to bus, ledger, and Telegram."""

from __future__ import annotations

import json
import logging
from datetime import UTC, datetime

from bond_monitor.domain.notifications.fingerprint import (
    alert_fingerprint,
    should_send_telegram,
    telegram_allowed_for_alert,
)
from bond_monitor.domain.notifications.models import Alert
from bond_monitor.domain.notifications.policies import DEFAULT_NOTIFICATION_POLICY, NotificationPolicy
from bond_monitor.infrastructure.notifications.ledger_repository import LedgerRepository
from bond_monitor.infrastructure.notifications.notifications_repository import NotificationsRepository
from bond_monitor.infrastructure.notifications.redis_bus import NotificationBus
from bond_monitor.infrastructure.notifications.telegram_client import TelegramNotifier

logger = logging.getLogger(__name__)


def format_telegram_message(alert: Alert, portfolio_name: str) -> str:
    prefix = "🔴" if alert.urgency == "critical" else "🟡"
    return (
        f"{prefix} {portfolio_name}\n"
        f"{alert.name} ({alert.isin})\n"
        f"{alert.reason}"
    )


class DeliverNotificationsUseCase:
    def __init__(
        self,
        *,
        ledger: LedgerRepository,
        bus: NotificationBus | None,
        telegram: TelegramNotifier,
        notifications_repo: NotificationsRepository | None = None,
        policy: NotificationPolicy | None = None,
    ) -> None:
        self._ledger = ledger
        self._bus = bus
        self._telegram = telegram
        self._notifications_repo = notifications_repo
        self._policy = policy

    @property
    def _effective_policy(self) -> NotificationPolicy:
        return self._policy or DEFAULT_NOTIFICATION_POLICY

    async def process_alert(self, alert: Alert, *, portfolio_name: str) -> None:
        fingerprint = alert_fingerprint(alert)
        self._ledger.ensure_detected(fingerprint, alert)
        await self._publish_bus(fingerprint, alert)
        await self._send_telegram_if_needed(fingerprint, alert, portfolio_name=portfolio_name)

    async def retry_pending(self, *, portfolio_names: dict[str, str]) -> None:
        for entry in self._ledger.list_pending_bus():
            payload = json.loads(entry.payload_json)
            alert = _alert_from_payload(payload)
            await self._publish_bus(entry.fingerprint, alert)
        for entry in self._ledger.list_pending_telegram():
            payload = json.loads(entry.payload_json)
            alert = _alert_from_payload(payload)
            portfolio_name = portfolio_names.get(alert.portfolio_id, alert.portfolio_id)
            await self._send_telegram_if_needed(entry.fingerprint, alert, portfolio_name=portfolio_name)

    async def _publish_bus(self, fingerprint: str, alert: Alert) -> None:
        entry = self._ledger.get(fingerprint)
        if entry is not None and entry.bus_published_at is not None:
            return
        now = datetime.now(UTC)
        published = False
        if self._bus is not None:
            try:
                self._bus.publish(
                    fingerprint=fingerprint,
                    portfolio_id=alert.portfolio_id,
                    kind=alert.kind.value,
                    payload=alert.to_payload(),
                    urgency=alert.urgency,
                )
                published = True
            except Exception:
                logger.exception("Redis publish failed for %s", fingerprint)
        if not published and self._notifications_repo is not None:
            await self._notifications_repo.upsert_from_bus(
                fingerprint=fingerprint,
                portfolio_id=alert.portfolio_id,
                kind=alert.kind.value,
                payload=alert.to_payload(),
                urgency=alert.urgency,
                created_at=now,
            )
            published = True
        if published:
            self._ledger.mark_bus_published(fingerprint, at=now)

    async def _send_telegram_if_needed(
        self,
        fingerprint: str,
        alert: Alert,
        *,
        portfolio_name: str,
    ) -> None:
        if not telegram_allowed_for_alert(alert, policy=self._effective_policy):
            return
        entry = self._ledger.get(fingerprint)
        if entry is not None and entry.telegram_sent_at is not None:
            if not should_send_telegram(
                last_telegram_sent_at=entry.telegram_sent_at,
                policy=self._effective_policy,
            ):
                return
        if not self._telegram.configured:
            return
        text = format_telegram_message(alert, portfolio_name)
        if self._telegram.send_message(text):
            self._ledger.mark_telegram_sent(fingerprint, at=datetime.now(UTC))


def _alert_from_payload(payload: dict) -> Alert:
    from bond_monitor.domain.notifications.models import AlertKind

    return Alert(
        portfolio_id=str(payload["portfolio_id"]),
        kind=AlertKind(str(payload["kind"])),
        isin=str(payload["isin"]),
        name=str(payload["name"]),
        lots=int(payload["lots"]),
        figi=payload.get("figi"),
        reason=str(payload["reason"]),
        urgency=payload["urgency"],
        detail_key=str(payload["detail_key"]),
        due_date=_parse_date(payload.get("due_date")),
        chat_template=payload.get("chat_template"),
        suggested_price_pct=payload.get("suggested_price_pct"),
        market_price_pct=payload.get("market_price_pct"),
        risk_acknowledgeable=bool(payload.get("risk_acknowledgeable", False)),
        offer_window_status=payload.get("offer_window_status"),
        submission_start=_parse_date(payload.get("submission_start")),
        submission_end=_parse_date(payload.get("submission_end")),
        escalation_kinds=tuple(payload.get("escalation_kinds") or ()),
    )


def _parse_date(value: str | None):
    from datetime import date

    if not value:
        return None
    return date.fromisoformat(value)

File: backend/internal/application/notifications/deliver_use_case.go
package notifications

import (
	"context"
	"encoding/json"
	"time"

	domain "github.com/tonatos/instrumenta/backend/internal/domain/notifications"
	"github.com/tonatos/instrumenta/backend/internal/infrastructure/notifications"
	"github.com/tonatos/instrumenta/backend/internal/infrastructure/persistence"
)

// TelegramDeliveryGate decides whether the owner may receive Telegram push.
type TelegramDeliveryGate interface {
	CanReceiveTelegram(ctx context.Context, ownerTelegramID int64) bool
}

// DeliverUseCase publishes alerts to bus, ledger, and Telegram.
type DeliverUseCase struct {
	ledger   *notifications.LedgerRepository
	bus      *notifications.RedisBus
	telegram *notifications.TelegramClient
	repo     *persistence.NotificationsRepository
	gate     TelegramDeliveryGate
	policy   domain.NotificationPolicy
}

func NewDeliverUseCase(
	ledger *notifications.LedgerRepository,
	bus *notifications.RedisBus,
	telegram *notifications.TelegramClient,
	repo *persistence.NotificationsRepository,
	gate TelegramDeliveryGate,
) *DeliverUseCase {
	return &DeliverUseCase{
		ledger: ledger, bus: bus, telegram: telegram, repo: repo, gate: gate,
		policy: domain.DefaultNotificationPolicy,
	}
}

func (d *DeliverUseCase) ProcessAlert(ctx context.Context, alert domain.Alert, portfolioName string, ownerTelegramID int64) error {
	fingerprint := domain.AlertFingerprint(alert)
	if _, err := d.ledger.EnsureDetected(fingerprint, alert); err != nil {
		return err
	}
	if err := d.publishBus(ctx, fingerprint, alert); err != nil {
		return err
	}
	d.sendTelegramIfNeeded(ctx, fingerprint, alert, portfolioName, ownerTelegramID)
	return nil
}

func (d *DeliverUseCase) RetryPending(ctx context.Context, portfolioNames map[string]string, ownerByPortfolio map[string]int64) error {
	pendingBus, err := d.ledger.ListPendingBus()
	if err != nil {
		return err
	}
	for _, entry := range pendingBus {
		alert, err := alertFromLedger(entry)
		if err != nil {
			continue
		}
		_ = d.publishBus(ctx, entry.Fingerprint, alert)
	}
	pendingTG, err := d.ledger.ListPendingTelegram()
	if err != nil {
		return err
	}
	for _, entry := range pendingTG {
		alert, err := alertFromLedger(entry)
		if err != nil {
			continue
		}
		name := portfolioNames[alert.PortfolioID]
		if name == "" {
			name = alert.PortfolioID
		}
		d.sendTelegramIfNeeded(ctx, entry.Fingerprint, alert, name, ownerByPortfolio[alert.PortfolioID])
	}
	return nil
}

func (d *DeliverUseCase) publishBus(ctx context.Context, fingerprint string, alert domain.Alert) error {
	entry, _ := d.ledger.Get(fingerprint)
	if entry != nil && entry.BusPublishedAt != nil {
		return nil
	}
	now := time.Now().UTC()
	payload := alertPayload(alert)
	published := false
	if d.bus != nil {
		if _, err := d.bus.Publish(ctx, fingerprint, alert.PortfolioID, string(alert.Kind), payload, string(alert.Urgency)); err == nil {
			published = true
		}
	}
	if !published && d.repo != nil {
		if _, err := d.repo.UpsertFromBus(ctx, fingerprint, alert.PortfolioID, string(alert.Kind), payload, string(alert.Urgency), &now); err != nil {
			return err
		}
	}
	if published || d.repo != nil {
		_, _ = d.ledger.MarkBusPublished(fingerprint, now)
	}
	return nil
}

func (d *DeliverUseCase) sendTelegramIfNeeded(ctx context.Context, fingerprint string, alert domain.Alert, portfolioName string, ownerTelegramID int64) {
	if d.telegram == nil || !d.telegram.Configured() || !domain.TelegramAllowedForAlert(alert, d.policy) {
		return
	}
	if ownerTelegramID == 0 {
		return
	}
	if d.gate != nil && !d.gate.CanReceiveTelegram(ctx, ownerTelegramID) {
		return
	}
	entry, _ := d.ledger.Get(fingerprint)
	var lastSent *time.Time
	if entry != nil {
		lastSent = entry.TelegramSentAt
	}
	if !domain.ShouldSendTelegram(lastSent, d.policy, time.Now()) {
		return
	}
	prefix := "🟡"
	if alert.Urgency == domain.AlertUrgencyCritical {
		prefix = "🔴"
	}
	text := prefix + " " + portfolioName + "\n" + alert.Name + " (" + alert.ISIN + ")\n" + alert.Reason
	if d.telegram.SendToChat(ownerTelegramID, text) {
		_, _ = d.ledger.MarkTelegramSent(fingerprint, time.Now().UTC())
	}
}

func alertPayload(alert domain.Alert) map[string]any {
	payload := map[string]any{
		"portfolio_id": alert.PortfolioID,
		"kind":         string(alert.Kind),
		"isin":         alert.ISIN,
		"name":         alert.Name,
		"reason":       alert.Reason,
		"urgency":      string(alert.Urgency),
	}
	if alert.ExtraPayload != nil {
		for k, v := range alert.ExtraPayload {
			if _, exists := payload[k]; exists {
				continue
			}
			payload[k] = v
		}
	}
	if alert.FIGI != nil {
		payload["figi"] = *alert.FIGI
	}
	if alert.SuggestedPricePct != nil {
		payload["suggested_price_pct"] = *alert.SuggestedPricePct
	}
	if alert.MarketPricePct != nil {
		payload["market_price_pct"] = *alert.MarketPricePct
	}
	if alert.Lots > 0 {
		payload["lots"] = alert.Lots
	}
	if alert.DueDate != nil {
		payload["due_date"] = alert.DueDate.Format("2006-01-02")
	}
	return payload
}

func alertFromLedger(entry domain.LedgerEntry) (domain.Alert, error) {
	var payload map[string]any
	if err := json.Unmarshal([]byte(entry.PayloadJSON), &payload); err != nil {
		return domain.Alert{}, err
	}
	alert := domain.Alert{
		PortfolioID: str(payload["portfolio_id"]),
		Kind:        domain.AlertKind(str(payload["kind"])),
		ISIN:        str(payload["isin"]),
		Name:        str(payload["name"]),
		Reason:      str(payload["reason"]),
		Urgency:     domain.AlertUrgency(str(payload["urgency"])),
	}
	if figi, ok := payload["figi"].(string); ok {
		alert.FIGI = &figi
	}
	if due, ok := payload["due_date"].(string); ok {
		if t, err := time.Parse("2006-01-02", due); err == nil {
			alert.DueDate = &t
		}
	}
	return alert, nil
}

func str(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}
